File: app/src/main/java/com/letter/socketassistant/connection/UdpConnection.java
package com.letter.socketassistant.connection;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;

public class UdpConnection extends Thread {

    private String remoteIp;
    private int remotePort;
    private int localPort;
    private InetAddress remoteAddress;
    private DatagramSocket datagramSocket;
    private DatagramPacket packetTransmit;
    private DatagramPacket packetReceived;
    private ConnectionReceivedListener connectionReceivedListener;

    private byte[] data = new byte[1024];

    public UdpConnection (String remoteIp, int remotePort, int localPort) {
        this.remoteIp = remoteIp;
        this.remotePort = remotePort;
        this.localPort = localPort;
        try {
            remoteAddress = InetAddress.getByName(remoteIp);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        try {
            datagramSocket = new DatagramSocket(localPort);

            packetReceived = new DatagramPacket(data, data.length);

            while (true) {
                datagramSocket.receive(packetReceived);
                if (connectionReceivedListener != null) {
                    connectionReceivedListener.onReceivedListener(new String(data, 0, packetReceived.getLength()));
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        } finally {
            try {
                if (datagramSocket != null) {
                    datagramSocket.close();
                }
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        }
    }

    public void setConnectionReceivedListener(ConnectionReceivedListener connectionReceivedListener) {
        this.connectionReceivedListener = connectionReceivedListener;
    }

    public void write (String string) {
        try {
            packetTransmit = new DatagramPacket(string.getBytes(), string.length(), remoteAddress, remotePort);
            datagramSocket.send(packetTransmit);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void disconnect () {
        try {
            if (datagramSocket != null) {
                datagramSocket.close();
            }
        } catch (Exception e1) {
            e1.printStackTrace();
        }
    }
}

File: app/src/main/java/com/letter/socketassistant/connection/UdpConnection.kt
package com.letter.socketassistant.connection

import android.util.Log
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress

/**
 * UDP 连接
 * @property remoteIp String 远端IP
 * @property remotePort Int 远端端口
 * @property maxPacketLen Int 接收最大包长
 * @property socket DatagramSocket socket连接实例
 * @constructor 构造一个UDP连接
 *
 * @author Letter(nevermindzzzt@gmail.com)
 * @since 1.0.0
 */
class UdpConnection constructor(private var remoteIp: String?,
                                private var remotePort: Int,
                                localPort: Int,
                                private val maxPacketLen: Int = 1024)
    : AbstractConnection() {

    companion object {
        private const val TAG = "UdpConnection"
    }

    private val socket = DatagramSocket(localPort)

    init {
        name = "udp: $remoteIp:$remotePort"
    }

    override fun send(connection: AbstractConnection, bytes: ByteArray?) {
        try {
            socket.send(DatagramPacket(
                bytes,
                bytes?.size ?: 0,
                InetAddress.getByName(remoteIp),
                remotePort
            ))
        } catch (e : Exception) {
            Log.e(TAG, "", e)
        }
    }

    override fun run() {
        val data = ByteArray(maxPacketLen)
        val packetReceived = DatagramPacket(data, data.size)
        while (!isInterrupted) {
            try {
                socket.receive(packetReceived)
                onReceivedListener?.invoke(this, data.sliceArray(IntRange(0, packetReceived.length - 1)))
            } catch (e : Exception) {
                Log.w(TAG, "", e)
                break
            }
        }
        onDisConnectedListener?.invoke(this)
    }

    override fun disconnect() {
        socket.close()
        super.disconnect()
    }
}